hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/scheduler/ContainerScheduler.java

```diff
@@ -0,0 +1,393 @@
+          "Container Killed to make room for Guaranteed Container.");
+      oppContainersMarkedForKill.put(contToKill.getContainerId(), contToKill);
+      LOG.info(
+          "Opportunistic container {} will be killed in order to start the "
+              + "execution of guaranteed container {}.",
+          contToKill.getContainerId(), container.getContainerId());
+    }
+  }
+
+  private void startAllocatedContainer(Container container) {
+    LOG.info("Starting container [" + container.getContainerId()+ "]");
+    scheduledToRunContainers.put(container.getContainerId(), container);
+    this.utilizationManager.addContainerResources(container);
+    if (container.getContainerTokenIdentifier().getExecutionType() ==
+        ExecutionType.OPPORTUNISTIC) {
+      this.opportunisticContainersStatus.setOpportMemoryUsed(
+          this.opportunisticContainersStatus.getOpportMemoryUsed()
+              + container.getResource().getMemorySize());
+      this.opportunisticContainersStatus.setOpportCoresUsed(
+          this.opportunisticContainersStatus.getOpportCoresUsed()
+              + container.getResource().getVirtualCores());
+      this.opportunisticContainersStatus.setRunningOpportContainers(
+          this.opportunisticContainersStatus.getRunningOpportContainers()
+              + 1);
+    }
+    container.sendLaunchEvent();
+  }
+
+  private List<Container> pickOpportunisticContainersToKill(
+      ContainerId containerToStartId) {
+    // The additional opportunistic containers that need to be killed for the
+    // given container to start.
+    List<Container> extraOpportContainersToKill = new ArrayList<>();
+    // Track resources that need to be freed.
+    ResourceUtilization resourcesToFreeUp = resourcesToFreeUp(
+        containerToStartId);
+
+    // Go over the running opportunistic containers.
+    // Use a descending iterator to kill more recently started containers.
+    Iterator<Container> reverseContainerIterator =
+        new LinkedList<>(scheduledToRunContainers.values()).descendingIterator();
+    while(reverseContainerIterator.hasNext() &&
+        !hasSufficientResources(resourcesToFreeUp)) {
+      Container runningCont = reverseContainerIterator.next();
+      if (runningCont.getContainerTokenIdentifier().getExecutionType() ==
+          ExecutionType.OPPORTUNISTIC) {
+
+        if (oppContainersMarkedForKill.containsKey(
+            runningCont.getContainerId())) {
+          // These containers have already been marked to be killed.
+          // So exclude them..
+          continue;
+        }
+        extraOpportContainersToKill.add(runningCont);
+        ResourceUtilizationManager.decreaseResourceUtilization(
+            getContainersMonitor(), resourcesToFreeUp,
+            runningCont.getResource());
+      }
+    }
+    if (!hasSufficientResources(resourcesToFreeUp)) {
+      LOG.warn("There are no sufficient resources to start guaranteed [{}]" +
```

Comment:
Actually that message might be a bit misleading. It should be re-worded to 'There are no sufficient resources at the moment to start guaranteed...'. You can only reach here if kill signals have been sent to Opp. containers, but the completed container event hasn't been received by the Scheduler yet. At which point the guaranteed container will start.
It is never possible to allocate a guaranteed container on node that cannot start it.